Better airdrop production

penev92 opened the pull request:
Due to GitHub being even more retarded than usual, I have to file a new PR instead of reopening #7468.
This is the same thing + the fixes.
The original C&C Nod Airstrip behavior is in place, as described in @pchote's last [comment](https://github.com/OpenRA/OpenRA/pull/7468#issuecomment-84653416).
Only thing left is to decide whether D2k carryalls should be delivered by the space frigate, or spawned on the map's edge when the frigate arrives, like the original game did.

Also doesn't contain the change regarding available subcells that was done in #7468.

penev92 committed:
Enable production traits to produce a collection of actors at once as foundation for bulk production,

Refactor Production and ProductionFromMapEdge traits,
Move Production.cs, ProductionAirdrop.cs, ProductionQueue.cs and ProductionQueueFromSelection.cs to the appropriate folders

Fixup: allow deliveries by both helicopters and planes

Phrohdoh commented on OpenRA.Mods.Common/Traits/Production/Production.cs:
```diff
@@ -99,27 +101,37 @@ public void DoProduction(Actor self, ActorInfo producee, ExitInfo exitinfo, stri
+			return allProduced;
+		}
+
+	    public static ExitInfo GetAvailableExit(Actor self, ActorInfo actorToProduce)
```
This indentation is wrong.

Phrohdoh commented on OpenRA.Mods.Common/Traits/Production/ProductionAirdrop.cs:
```diff
@@ -0,0 +1,200 @@
+
+		public override bool Produce(Actor self, IEnumerable<ActorInfo> actorsToProduce, string raceVariant)
+		{
+			//var info = (ProductionAirdropInfo)Info;
```
If we don't need this get rid of it.

penev92 commented:
Updated. I still have no idea what Travis's problem is.

Mailaender commented:
You forgot to delete https://github.com/penev92/OpenRA/blob/bleed_betterAirdropProduction/OpenRA.Mods.Cnc/Traits/Buildings/ProductionAirdrop.cs

penev92 commented:
Finally fixed the build.

Phrohdoh commented on OpenRA.Mods.Common/Traits/Production/ProductionAirdrop.cs:
```diff
@@ -0,0 +1,199 @@
+				new CenterPositionInit(spawn)
+			};
+
+			if (raceVariant != null)
```
You may want `string.IsNullOr(Empty|Whitespace)`.

penev92 replied:
Good point.

Phrohdoh commented on OpenRA.Mods.Common/Traits/Production/ProductionAirdrop.cs:
```diff
@@ -0,0 +1,199 @@
+			this.self = self;
+			factionVariant = raceVariant;
+
+			// Check if there is a valid drop-off point before sending the transport
```
I think this is explained by the code.

Phrohdoh commented on OpenRA.Mods.Common/Traits/Production/ProductionAirdrop.cs:
```diff
@@ -0,0 +1,199 @@
+			if (exit == null)
+				return false;
+
+			foreach (var tower in self.TraitsImplementing<INotifyDelivery>())
```
`tower` feels out of place here.
I get that what you're relating it to but I don't think it has a place as a variable name here.

penev92 replied:
Yeah, that is a leftover from the old code, probably from before it was intended to use for anything other than the TD Nod Airstrip.

penev92 committed:
Add an upgrade rule for ProductionAirdrop delivering actor and renaming the trait

penev92 commented:
Updated.

Mailaender commented:
Can you give a quick summary what has been improved here? I just noticed that space frigates can now get stuck if the landing pad exit is blocked.

penev92 commented:
The most noteworthy point, I think, is:
 - The Production* traits get support for producing a collection of actors, instead of only the one. This functionality will be utilized in the follow-up PR.

Other important points are:
 - The TD Nod delivery plane's behavior now matches the original game in the cases where there is no space to drop a new unit.
 - In both TD and D2k the delivery actors will wait until all cargo is unloaded, as opposed to dropping everything off immediately on touchdown and taking off again.

Small fries:
 - As a side bonus, D2k uses one less trait from `OpenRA.Mods.Cnc`.
 - There was a bit of Production* trait reordering (added a new `Traits\Production\` folder).
 - Some cosmetic fixes to `Production` and `ProductionFromMapEdge`.

penev92 commented:
Yet another pointless rebase happened, with the usual merge conflicts in `UpgradeRules.cs` (but this time Git decided to make things more interesting and screwed up an auto-merge of another file, without so much as a warning).
I wonder how many more times I'll have to rebase this.

abcdefg30 commented on OpenRA.Mods.D2k/Traits/Buildings/ProductionFromMapEdge.cs:
```diff
@@ -24,59 +25,70 @@ class ProductionFromMapEdgeInfo : ProductionInfo
 
-					if (raceVariant != null)
-						td.Add(new RaceInit(raceVariant));
+			if (raceVariant != null)
```
Why not the same as [above](https://github.com/OpenRA/OpenRA/pull/7786/files#diff-3a1930006db09a5bddaed5ff0fb96974R69)?

Phrohdoh replied:
Ought to use `string.IsNullOrEmpty` for extra safety.

penev92 replied:
Done.

pchote commented on OpenRA.Mods.Common/Traits/Production/ProductionByDelivery.cs:
```diff
@@ -0,0 +1,198 @@
+	{
+		[Desc("Delivering actor name.")]
+		[ActorReference]
+		public readonly string DeliveringActorType = "c17";
```
I don't think we use `Type` in ActorReference names anywhere else.

pchote replied:
I suggest `DeliveryActor` or `Transport`.

penev92 commented:
Renamed the space-faring C17 to be `DeliveryActor`.

obrakmann commented on OpenRA.Mods.Common/Traits/Production/Production.cs:
```diff
@@ -37,48 +38,49 @@ public class Production
+				td.Add(new RaceInit(raceVariant));
 
-				var newUnit = self.World.CreateActor(producee.Name, td);
+			var newUnit = self.World.CreateActor(false, actorToProduce.Name, td);
```
I believe this will make actors tick a tick earlier than before, but I haven't noticed any impact from that change.

penev92 replied:
The actors are created at the same moment as before. They are just not added to the world until at least the end of the frame, or until there is a free exit they can use.
They shouldn't be ticking if they aren't in the world, so if anything, this *should* make them tick later.

obrakmann replied:
The actors that aren't in the world don't tick *themselves*, so they don't run any activities, but their traits *do* tick. At least that's how I read [World::Tick()](https://github.com/OpenRA/OpenRA/blob/bleed/OpenRA.Game/World.cs#L292-L295).

Also I'm not sure how the `TraitDict` handles being modified while being iterated (but it seems to work).

obrakmann commented on OpenRA.Mods.Common/Traits/Production/ProductionByDelivery.cs:
```diff
@@ -0,0 +1,198 @@
+			// Start a fixed distance away: the width of the map.
+			// This makes the production timing independent of spawnpoint
+			startPos = self.Location + new CVec(self.Owner.World.Map.Bounds.Width, 0);
+			endPos = new CPos(self.Owner.World.Map.Bounds.Left - 5, self.Location.Y);
```
Why 5?

pchote replied:
This is copied from my original `ProductionAirdrop`.  I picked 5 so that the plane moves completely outside the map before disappearing.

Mailaender replied:
In that case we will need to revive https://github.com/OpenRA/OpenRA/pull/6874 as the minimal 1 cell map cordon won't be enough then.

penev92 replied:
Aircraft don't care about cordons, and while I tried making the trait as generic as possible, it's still limited to delivery by aircraft.

obrakmann commented on OpenRA.Mods.Common/Traits/Production/ProductionByDelivery.cs:
```diff
@@ -0,0 +1,198 @@
+
+			owner.World.AddFrameEndTask(w =>
+			{
+				var altitude = self.World.Map.Rules.Actors[deliveringActorType].Traits.Get<PlaneInfo>().CruiseAltitude;
```
How about `AircraftInfo` instead? Then people can use helicopters for delivery as well.

pchote replied:
We can branch of helicopter/plane and then queue the appropriate flight activities.

obrakmann commented on OpenRA.Mods.Common/Traits/Production/ProductionByDelivery.cs:
```diff
@@ -0,0 +1,198 @@
+	[Desc("Deliver produced units from outside the map.")]
+	public class ProductionByDeliveryInfo : ProductionInfo
+	{
+		[Desc("Delivering actor name.")]
```
You should mention that that actor needs to be `Plane`.

pchote replied:
I like the idea of instead fixing this to support helicopters (but then still fixing this to mention it needs to be Helicopter of Plane).  Helicopter flight physics would fit the d2k delivery better than Plane.

obrakmann replied:
I just looked up the original frigate's behaviour on youtube, and I have to agree. It's much closer to our helicopter behaviour. So it would be nice if this could work with both planes and helicopters. It'll make the code a bit uglier, but be nicer in-game.

penev92 replied:
There are a ton more stuff that need to be changed before we have a proper starport, and I am not inclined to do everything in 1 PR. The original PRs were 3 (@WolfGaming had 2 and I had a follow-up of those), and I've since stated a few times that there will be *at least one* PR after this one.

obrakmann replied:
For some reason I can't open a PR against your repo, it doesn't show up in the dropdown list.

I'm afraid you'll have to apply that manually then: https://github.com/OpenRA/OpenRA/commit/5802b7384db654b23231ed045d4369f191797909 ([raw diff for download](https://github.com/OpenRA/OpenRA/commit/5802b7384db654b23231ed045d4369f191797909.diff))

pchote commented on OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs:
```diff
@@ -943,6 +943,21 @@ internal static void UpgradeActorRules(int engineVersion, ref List<MiniYamlNode>
 				}
 
+				// Rename ProductionAirdrop delivering ActorType and rename the trait
+				if (engineVersion < 20150508)
```
Don't forget to update this once we are ready to merge.

obrakmann commented on OpenRA.Mods.Common/Traits/Production/ProductionByDelivery.cs:
```diff
@@ -0,0 +1,198 @@
+				var value = actorsToProduce.Sum(actor => actor.Traits.Get<ValuedInfo>().Cost);
+				self.Owner.PlayerActor.Trait<PlayerResources>().GiveCash(value);
+
+				return;
```
I may have missed it, but I think the `deliveringActor` doesn't get removed from the world in this case, does it?

obrakmann commented:
Overall the code looks ok and works as expected in-game.

So basically just three changes needed to get my :+1::
- [ ] Remove the delivering actor properly from the world when there are no exits
- [ ] Support both planes and helicopters
- [ ] Mention that fact in the `[Desc()]`

obrakmann commented on OpenRA.Mods.Common/Traits/Production/ProductionByDelivery.cs:
```diff
@@ -0,0 +1,198 @@
+			var deliveringActorType = info.DeliveryActor;
+			var owner = self.Owner;
+
+			this.self = self;
```
Wait... why do you re-assign `self` here?

penev92 commented:
Rebased and merged @obrakmann's commit, which should cover the current set of requested changes.

obrakmann commented:
It would be nice if you could squash my commit into your first one. It wasn't really meant to stand alone (hence the 'fixup' in the title ;))

abcdefg30 commented:
Got this crash:
```crash
Tiberian Dawn Mod at Version {DEV_VERSION}
Operating System: Windows (Microsoft Windows NT 6.2.9200.0)
Runtime Version: .NET CLR 4.0.30319.34209
Exception of type `System.InvalidOperationException`: Attempted to get trait from destroyed object (afld 74 (not in world))
   at OpenRA.TraitDictionary.CheckDestroyed(Actor actor) in \OpenRA\OpenRA.Game\TraitDictionary.cs:Line 84.
   at OpenRA.TraitDictionary.GetOrDefault[T](Actor actor) in \OpenRA\OpenRA.Game\TraitDictionary.cs:Line 101.
   at OpenRA.Actor.TraitOrDefault[T]() in \OpenRA\OpenRA.Game\Actor.cs:Line 200.
   at OpenRA.Traits.Target.FromActor(Actor a) in \OpenRA\OpenRA.Game\Traits\Target.cs:Line 48.
   at OpenRA.Mods.Common.Traits.ProductionByDelivery.TryToLand(Actor deliveringActor, IEnumerable`1 actorsToProduce)
   at OpenRA.Mods.Common.Traits.ProductionByDelivery.<>c__DisplayClass4.<>c__DisplayClass6.<Produce>b__3()
   at OpenRA.Activities.CallFunc.Tick(Actor self) in \OpenRA\OpenRA.Game\Activities\CallFunc.cs:Line 29.
   at OpenRA.Traits.Util.RunActivity(Actor self, Activity act) in \OpenRA\OpenRA.Game\Traits\Util.cs:Line 83.
   at OpenRA.Actor.Tick() in \OpenRA\OpenRA.Game\Actor.cs:Line 121.
   at OpenRA.World.Tick() in \OpenRA\OpenRA.Game\World.cs:Line 293.
   at OpenRA.Game.InnerLogicTick(OrderManager orderManager) in \OpenRA\OpenRA.Game\Game.cs:Line 461.
   at OpenRA.Game.LogicTick() in \OpenRA\OpenRA.Game\Game.cs:Line 485.
   at OpenRA.Game.Loop() in \OpenRA\OpenRA.Game\Game.cs:Line 601.
   at OpenRA.Game.Run() in \OpenRA\OpenRA.Game\Game.cs:Line 641.
   at OpenRA.Program.Run(String[] args) in \OpenRA\OpenRA.Game\Support\Program.cs:Line 111.
   at OpenRA.Program.Main(String[] args) in \OpenRA\OpenRA.Game\Support\Program.cs:Line 39.
```

Steps to reproduce:
1) Build an airstrip and produce a unit.
2) Sell the airstrip before the c17 plane arrives (enters the visible map area).
3) Enjoy your crash. :(

penev92 commented:
Fixed the crash and squashed.

pchote commented on OpenRA.Mods.Common/Scripting/Properties/ProductionProperties.cs:
```diff
@@ -39,7 +39,7 @@ public void Produce(string actorType, string raceVariant = null)
 				throw new LuaException("Unknown actor type '{0}'".F(actorType));
 
-			Self.QueueActivity(new WaitFor(() => p.Produce(Self, actorInfo, raceVariant)));
+			Self.QueueActivity(new WaitFor(() => p.Produce(Self, new[] { actorInfo }, raceVariant)));
```
This will create a new array each time WaitFor ticks.  Please extract this out into a temporary variable outside the closure.

pchote commented on OpenRA.Mods.Common/Traits/Player/ClassicProductionQueue.cs:
```diff
@@ -107,11 +107,11 @@ protected override bool BuildUnit(string name)
-					return true;
-				}
+				if (!p.Trait.Produce(p.Actor, new[] { ai }, p.Trait.Race))
+					continue;
```
I find this change in flow control to be less clear than what it had originally.

pchote commented on OpenRA.Mods.Common/Traits/Production/Production.cs:
```diff
@@ -102,27 +104,37 @@ public void DoProduction(Actor self, ActorInfo producee, ExitInfo exitinfo, stri
 			{
-				DoProduction(self, producee, exit, raceVariant);
-				return true;
+				// pick a spawn/exit point pair
```
Minor nit: ``// Pick`

pchote commented on OpenRA.Mods.Common/Traits/Production/ProductionByDelivery.cs:
```diff
@@ -0,0 +1,223 @@
+			isPlane = aircraft is PlaneInfo;
+		}
+
+		public override Actor DoProduction(Actor self, ActorInfo actorToProduce, ExitInfo exitInfo, string raceVariant)
```
It would be really good if we didn't have to override the base implementation to make this simple again.  Could you extract the extra stuff that the base method does into the other outer DoProduction call so that this code can be shared?

pchote commented on OpenRA.Mods.Common/Traits/Production/ProductionByDelivery.cs:
```diff
@@ -0,0 +1,223 @@
+			deliveringActor.QueueActivity(new CallFunc(() => MakeDelivery(actorsToProduce.ToList(), deliveringActor, deliveringActor.World)));
+		}
+
+		void MakeDelivery(ICollection<ActorInfo> actorInfos, Actor deliveringActor, World world)
```
What does `MakeDelivery` mean?

pchote replied:
It would be clearer to declare `List<ActorInfo> actorInfos` here and in `WaitAndRetry`.

pchote commented on OpenRA.Mods.Common/Traits/Production/ProductionByDelivery.cs:
```diff
@@ -0,0 +1,223 @@
+			}
+		}
+
+		void WaitAndRetry(Actor actor, World world, ICollection<ActorInfo> actorInfos)
```
This is only called once from inside `MakeDelivery`.  I think it would be clearer to not extract this as a method.

pchote commented on OpenRA.Mods.Common/Traits/Production/ProductionByDelivery.cs:
```diff
@@ -0,0 +1,223 @@
+			actor.QueueActivity(new CallFunc(() => MakeDelivery(actorInfos, actor, world)));
+		}
+
+		static void IssueNotifications(Actor self, Actor newActor, CPos exitLocation)
```
Could you move this in to Production (and then maybe rename it to something like `AfterProduction`) to reduce duplication?

pchote commented on OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs:
```diff
@@ -943,6 +943,21 @@ internal static void UpgradeActorRules(int engineVersion, ref List<MiniYamlNode>
+						if (child != null)
+							node.Value.Nodes.Add(new MiniYamlNode("DeliveryActor", child.Value.Value));
+
+						node.Value.Nodes.RemoveAll(n => n.Key == "ActorType");
```
This is a very complicated way to rename a key.  Look up to see how other upgrade rules do it by reassigning n.Key.

pchote commented:
If you spam production on the Nod airstrip, then the first 5 or so planes will all land on top of eachother on the strip, and the remaining planes will continue flying without landing.  Can you use the reservation mechanic so that only the first plane will land?

pchote commented:
For the sake of not breaking debug mode production, could the production trait hold a reference to the last spawned production aircraft, and then use that same plane for all production in the following N ticks (whatever time is required for a plane to land and unload)?  This would fix the problem where spamming the production in debug mode will only give you each Nth actor.

penev92 commented:
- The current behavior works perfectly well with our TD mod.
 - It is an exact copy of the behaviour in the original C&C (as described [here](https://github.com/OpenRA/OpenRA/pull/7468#issuecomment-84653416))
 - The debug instabuild was "broken" because we wanted the original behavior (also I don't find this to be a problem).
 - If someone is making a mod and wants a C17 to deliver actors with < 3 second build times (that's when the problem will occur), they really need to rethink their delivery method, because that makes no sense.
 - Adding newly produced tanks onto a plane that is already on the map and flying toward the airstrip is just a different kind of bogus.

pchote commented:
Breaking the debug mode isn't acceptable, so I have to :-1: this until it can be fixed.  The general approach I have suggested will not affect the behaviour for regular gameplay, and in all but the most extreme case (1 cell map cordon, airstrip placed on the very left hand side of the map, very fast aircraft) would add the actors to the C17 before it gets anywhere near the playable world.

pchote commented on OpenRA.Mods.Common/Traits/Production/ProductionByDelivery.cs:
```diff
@@ -0,0 +1,223 @@
+		public readonly string DeliveryActor = "c17";
+
+		public readonly string ReadyAudio = "Reinforce";
+
```
Add a `public readonly int MinimumInterval = 125;` here.

pchote commented on OpenRA.Mods.Common/Traits/Production/ProductionByDelivery.cs:
```diff
@@ -0,0 +1,223 @@
+		}
+
+		public override bool Produce(Actor self, IEnumerable<ActorInfo> actorsToProduce, string raceVariant)
+		{
```
Make this add `actorsToProduce` to a collection and then run the rest of this (with the aggregate collection) from inside a `Tick()` method every `MinimumInterval` ticks.

pchote commented:
The two latest inline comments suggest an even simpler fix to the debug issue, and can then be extended later for the starport.

penev92 commented:
>[17:48] <penev> I'm not inclined to add a pile of redundant code that does something very questionable ingame
[17:48] <penev> just so that it dan work in debug in a different way than normal
[17:49] <penev> also as I mentioned before, I had lost almost any interest in updating that by the time antares sent me his fixup commit
[17:49] <penev> and I have no interest in arguing over how it works in debug now
[17:50] <penev> also that thing has been open for at least 3 monts
[17:50] <penev> with 2 reviews so far
[17:51] <penev> and finally, I may end up having to split off the starport logic anyway
[17:51] <penev> so I'm just going to leave this as it is on bleed

penev92 committed:
Code fixes

penev92 committed:
"Fix" instabuild

penev92 committed:
Fix travis

abcdefg30 commented:
Is this intentional?
![starport](https://cloud.githubusercontent.com/assets/7704140/7779791/ccd73076-00d7-11e5-9be4-36273a972372.png)
(As you can see here, the frigate landed on a mcv.)

I'd also vote for the carryalls coming alone, not in a frigate.

Looks otherwise good to me. :+1: / :white_check_mark: when above is resolved and the commits are squashed.

Mailaender commented:
Needs a rebase.

penev92 commented:
Rebased.

Mailaender commented:
Needs a rebase again.

penev92 commented:
Rebased.

pchote commented on OpenRA.Mods.Common/Traits/Production/ProductionByDelivery.cs:
```diff
@@ -0,0 +1,190 @@
+		public override bool Produce(Actor self, IEnumerable<ActorInfo> actorsToProduce, string raceVariant)
+		{
+			production.AddRange(actorsToProduce);
+			factionVariant = raceVariant;
```
It would be a bit more robust if you could store this together with each production item.

penev92 committed:
Make race variant per-actor

penev92 commented:
Rebased, updated the upgrade rule date and added a new commit that makes it possible to produce N actors at once with a different race variant for each.
Once the new commits are deemed OK I'll squash them before merge.

pchote commented:
I found two issues while testing:
* If the afld is sold or destroyed while the plane is waiting to unload it should take off (or maybe blow up if destroyed?) and refund the remaining units instead of sitting on the ground.
* I am able to trigger a situation where a plane never arrives, and the afld sits there playing the waiting-for-delivery animation forever.  Not sure how, though.

penev92 committed:
Don't get stuck on the ground!

penev92 commented:
Implemented the delivery actor flying away/dieing (configurable through YAML) when the airfield/starport is destroyed/sold.

Mailaender commented:
This looks a lot better for d2k. I noticed a glitch when I sell the starport during approach where it lands a few cells before the starport but the unload beams it to the location of the exit.

matija-hustic commented:
I've noticed a couple of things:
* The waiting-for-delivery animation never gets stopped after delivery (`INotifyDelivery.Delivered` never called).
* When selling or destroying the starport (probably the nod airstrip, too) after a successful call to `TryToLand`, but before `MakeDelivery` gets called and `landedActor` set, the delivery aircraft lands and unloads instead of immediately taking off again (since `landedActor == null`, `TakeOff` never gets called in `Killed` and `Selling`).
* When all exits from the starport (pbbly also airstrip) are blocked, every new order gets paid for, but automatically ignored (`production`cleared in `StartDelivery`).
* In extreme debug cases (ordering as if there's no tomorrow), two delivery aircraft are allowed to land and unload simultaneously, making four tanks appear on two cells simultaneously (`GetAvailableExit` can return non-null in `TryToLand` even when one delivery aircraft is already unloading). Nothing crashes tho.

penev92 commented:
Seems like this got a few more things requested while waiting on a green light from @pchote.
I have exams starting in a couple of days and will most likely be unable to finish this in the next few weeks, so I'll close it again for now.
